### DISCUSSION TITLE
Transfer dynamic data without blocking render?

### DISCUSSION THREAD

**Discussion by Anshi701** [vsg-dev/VulkanSceneGraph#1121] [Q&A] [answered]
Hello! Then dynamic data is changing, copying to GPU memory is done in render thread. But is there any way to transfer data without blocking main render thread?

**Comment by robertosfield** [comment, marked as answer]
The design of vsg::TransferTask is meant to ensure that the data is robustly copied to the GPU and available when required, if you drop the synchronization then you can end up consuming data before it's been updated.

The design of TransferTask is also meant for modest amounts of data, it scales pretty well to pretty large amounts of data.

There is also multi-buffering done by TransferTask to ensure updates on the CPU and copied to the GPU data at the correct time, otherwise you can end up with values from the wrong frame being consumed on the GPU.

If you don't want the safety that TransferTask provides then you can implement your own copying of data to the GPU, all the code is there available.

What is the best solution for your application is something I can't advise upon as you have provide any details why you want data to transfer without blocking, or what the data is being used for etc.  It may be that you are trying to do something in a way that is entirely inappropriate.

**Reply by Anshi701** [reply]
Thanks for your reply! I want to update pretty huge amount of data, 1-2 GB while flying across planet surface. Surface is drawn by tiles. Right now, then i fly pretty fast and updating is done pretty every 5-10 frames, data trancfering is decreases fps pretty much. In my previous app with opengl, i was doing double buffering and swapping after transferring is done.

**Reply by robertosfield** [reply]
TransferTask double or triple buffers(depending on size of the swap chain)  on the CPU but not on the GPU side.  It's a general purpose class for dynamic data and it needs to handle a wide rang of usage cases without pushing complexity on end users.

For your particular usage case you may be able to make choices that suit your usage case, rolling your own data transfer class might be the best thing to do.   All the source code to TransferTask is available for review and it's built from other VSG components to achieve the desired result, so rolling your own needed mean coding in Vulkan directly.

**Reply by Anshi701** [reply]
Got it, thank you for answering!
